### PR TITLE
Fetch Playlists from channels

### LINKED ISSUE #174: [Feature Request] get list of Playlists created by a user
image

This would be an awesome feature to have in `youtube-search-python`.

### PR DESCRIPTION
**Pull request by fabi321**
Fixes #174

It is still flawed. i.e. it doesn't support endless scrolling.
Furthermore I am not sure if ChannelSearch should be renamed to
ChannelVideoSearch, to be a bit more precise.

### CODE REVIEW

**Comment by mytja** [comment]
Hello, and thanks for your contribution, but this is really not an optimal solution.
First things first, that's not what he's asking for in #174 . He wants to retrieve all playlists (correct me if I'm wrong). I wanted to use `Playlist` or `Channel` class for that, as it already contains pretty good parser for playlists.

Your implementation is simply put, too complex. It contains too many classes, for example:
1. Why do we even use this? We don't need 2 classes to do a job, that one could. https://github.com/alexmercerind/youtube-search-python/pull/177/files#diff-e6569ebd018d9d9636c3eee438c36de8d9ee42102b6af978ded4a893400b7ee3R139
2. This `sync` hard-coding here (https://github.com/alexmercerind/youtube-search-python/pull/177/files#diff-e6569ebd018d9d9636c3eee438c36de8d9ee42102b6af978ded4a893400b7ee3R130) is far from optimal (for `async` users)
3. Why do we need another RequestCore? We already have one that's working great. All of this parsing is not intended for RequestCore, but for your core class (eg. ChannelPlaylistSearchCore) - https://github.com/alexmercerind/youtube-search-python/pull/177/files#diff-e6569ebd018d9d9636c3eee438c36de8d9ee42102b6af978ded4a893400b7ee3R11

So that your PR doesn't go to waste, I'm going to build on top of your PR and afterwards merge it. (as far as I know, you need to enable "Allow maintainers to update code" or something similar on your PR, but I think it should be by default enabled)

**Comment by fabi321** [comment]
About using the `Playlist` Class: AFAIK it's about getting all playlists, and not their content. The latter can be done later. and about using `Channel`: I see no code for handling playlists there. And about fetching all playlists: that was my goal, although I just spotted a bug, that prevents this.

**Comment by fabi321** [comment]
It's just, that playlists are organized in shelves, which don't provide all playlists, but only some as a preview, that's why I parsed each shelf (that's a naming error, it should be called something like `SingleShelfPlaylistSearch`) separately. I do agree that the sync hardcoding is bad.

**Comment by mytja** [comment]
This should be great. Tell me what you think.

File: youtubesearchpython/core/channelsearch.py [hunk 1]
```diff
@@ -85,76 +95,3 @@ def result(self, mode: int = ResultMode.dict) -> Union[str, dict]:
...
-                tab_renderer = last_tab["tabRenderer"]
-                if 'content' in tab_renderer:
-                    self.response = tab_renderer["content"]["sectionListRenderer"]["contents"]
-                else:
-                    self.response = []
-        except:
-            raise Exception('ERROR: Could not parse YouTube response.')
-
-
-class ChannelPlaylistSearchCore(ChannelRequestCore):
```
Comment (mytja) [inline youtubesearchpython/core/channelsearch.py]: This class, I might have accidentally removed. This class was intended for searching playlists within channel (eg. with a query), right? I can't seem to find this feature anywhere in YouTube webapp. Can you maybe screenshot where exactly can you pick between playlists and videos, as I can only search by everything. If you provide me this, I'll be able to implement this class (or you may for that matter).

Comment (fabi321) [reply youtubesearchpython/core/channelsearch.py]: This was probably just an misunderstanding on my side.

File: youtubesearchpython/core/channel.py [hunk 2]
```diff
@@ -8,20 +8,36 @@
 from youtubesearchpython.core.componenthandler import getValue, getVideoId
 
 
-
 class ChannelCore(RequestCore):
-    def __init__(self, channelId: str):
+    result = {}
```
Comment (fabi321) [inline youtubesearchpython/core/channel.py:12]: It is wrong to write it here. This is global and not local. Consider the following
```python
pew = Channel('UC-lHJZR3Gqxm24_Vd_AJ5Yw', ChannelRequestType.playlists)
beet = Channel('UC2XdaAVUannpujzv32jcouQ', ChannelRequestType.playlists)
print(pew.result)
```
The expected result are playlists from pewdiepie, but the observer results are playlists from the beetles.
The correct was is to put this into the `__init__` function.

File: youtubesearchpython/core/channel.py [hunk 3]
```diff
@@ -48,28 +79,58 @@ def parse_response(self):
...
+        response = self.data.json()
+
+        self.continuation = None
+
+        response = getValue(response, ["onResponseReceivedActions", 0, "appendContinuationItemsAction", "continuationItems"])
+        for i in response:
+            if getValue(i, ["continuationItemRenderer"]):
+                self.continuation = getValue(i, ["continuationItemRenderer", "continuationEndpoint", "continuationCommand", "token"])
+                break
+            self.result["playlists"].append(self.playlist_parse(i))
```
Comment (fabi321) [inline youtubesearchpython/core/channel.py:109]: This doesn't work. This way, all further playlists only contain None.
```suggestion
            elif getValue(i, ['gridPlaylistRenderer']):
                self.result["playlists"].append(self.playlist_parse(getValue(i, ['gridPlaylistRenderer'])))
            # TODO: Handle other types like gridShowRenderer
```

Comment (mytja) [reply youtubesearchpython/core/channel.py:109]: Hmm, it seems like I haven't done proper testing & printed out final results. Interesting...

File: youtubesearchpython/core/channel.py (hunk 2 above)
Comment (mytja) [reply youtubesearchpython/core/channel.py]: :+1:

File: youtubesearchpython/core/channel.py [hunk 4]
```diff
@@ -8,20 +8,35 @@
...
+            self.data["continuation"] = self.continuation
+
+    def playlist_parse(self, i) -> dict:
+        return {
+            "id": getValue(i, ["playlistId"]),
+            "thumbnails": getValue(i, ["thumbnail", "thumbnails"]),
+            "title": getValue(i, ["title", "runs", 0, "text"]),
+            "videoCount": getValue(i, ["videoCountShortText", "simpleText"]),
+            "viewCount": getValue(i, ["v", 1, "simpleText"]),
+            "lastEdited": getValue(i, ["publishedTimeText", "simpleText"]),
```
Comment (fabi321) [inline youtubesearchpython/core/channel.py:38]: Have you actually seen these in a playlist response? I haven't so far

Comment (mytja) [reply youtubesearchpython/core/channel.py:38]: I've seen it in YouTube's JSON, but cannot confirm that I correctly copy-pasted the paths. I seem to have forgotten to remove `viewCount`. It's definitely invalid. Will fix everything tomorrow (or you may now, if you want to).

Comment (fabi321) [reply youtubesearchpython/core/channel.py]: I just confirmed that `lastEdited` actually appears, and removed `viewCount`

**Comment by fabi321** [comment]
I don't see any bugs anymore, so feel free to merge it.

**Comment by fabi321** [comment]
The current implementation only works for normal channels. "* - Topic" channels don't work.
In example the Beatles: The current implementation only returns all playlists that are visible in `Created Playlists` [here](https://www.youtube.com/channel/UC2XdaAVUannpujzv32jcouQ/playlists). It neither returns what pops up when actually clicking on `Created Playlists`, nor what is in `Albums & Singles`

**Comment by fabi321** [comment]
Maybe this could be moved into another PR, to allow the navigation by shelves. I could live with that

**Comment by mytja** [comment]
`lastEdited` is definitely in response, but some playlists don't have it for some reason. Have a look at [NCS](https://www.youtube.com/c/NoCopyrightSounds/playlists?view=1&sort=dd&shelf_id=0). You can see some playlists have last date of editing (under playlist thumbnail in webapp), while some don't (for some reason).

About `- Topic` playlists. Most of these playlists are for music (not all, but most). These `- Topic` channels were created by YouTube internally for YouTube Music. This library supports only YouTube, but if you want YouTube Music, check it out [here](https://github.com/sigma67/ytmusicapi). You see, if I go to gaming `- Topic` channel, let's say [Among Us](https://www.youtube.com/channel/UCn0QAnNB_vHr7Pry3Vo72lg) (this was the first one in `Gaming` section), you can see there are no playlists whatsoever. These channels are not usual channels and should not be treated the same way. For retrieving albums, I recommend `ytmusicapi`.

After all, I think this PR is ready for merging.

**Pull request merged**